File: Managed Source/ArsenalImageMounterCLISetup/MainModule.vb

''''' MainModule.vb
''''' Console driver setup application, for scripting and similar.
''''' 
''''' Copyright (c) 2012-2021, Arsenal Consulting, Inc. (d/b/a Arsenal Recon) <http://www.ArsenalRecon.com>
''''' This source code and API are available under the terms of the Affero General Public
''''' License v3.
'''''
''''' Please see LICENSE.txt for full license terms, including the availability of
''''' proprietary exceptions.
''''' Questions, comments, or requests for clarification: http://ArsenalRecon.com/contact/
'''''

Imports Arsenal.ImageMounter.DriverSetup
Imports Arsenal.ImageMounter.API
Imports Arsenal.ImageMounter.IO
Imports System.Windows.Forms
Imports System.ComponentModel
Imports System.Reflection
Imports Arsenal.ImageMounter

Public Module MainModule

    Private ReadOnly ownerWindow As New NativeFileIO.NativeWindowHandle(NativeFileIO.SafeNativeMethods.GetConsoleWindow())

    Public Function Main(ParamArray args As String()) As Integer

        Trace.Listeners.Add(New ConsoleTraceListener)

        Dim opMode As OpMode = OpMode.Status

        If args?.Length > 0 Then
            If args(0).Equals("/install", StringComparison.OrdinalIgnoreCase) Then
                opMode = OpMode.Install
            ElseIf args(0).Equals("/uninstall", StringComparison.OrdinalIgnoreCase) Then
                opMode = OpMode.Uninstall
            ElseIf args(0).Equals("/status", StringComparison.OrdinalIgnoreCase) Then
                opMode = OpMode.Status
            Else
                Trace.WriteLine($"Syntax: {Assembly.GetExecutingAssembly().GetName().Name} /install|/uninstall|/status")
            End If
        End If

        Try
            Return SetupOperation(opMode)

        Catch ex As Exception
            If TypeOf ex Is Win32Exception Then
                Trace.WriteLine($"Win32 error: {DirectCast(ex, Win32Exception).NativeErrorCode}")
            End If
            Trace.WriteLine(ex.ToString())
            Return -1

        End Try

    End Function

    Public Function SetupOperation(opMode As OpMode) As Integer

        Trace.WriteLine($"Kernel type: {Kernel}")
        Trace.WriteLine($"Kernel supports StorPort: {HasStorPort}")

        Select Case opMode

            Case OpMode.Install
                Using zipStream = GetType(MainModule).Assembly.GetManifestResourceStream(
                        GetType(MainModule), "DriverFiles.zip")

                    InstallFromZipStream(ownerWindow, zipStream)

                End Using

                Try
                    Using New ScsiAdapter
                    End Using

                    Trace.WriteLine("Driver successfully installed.")
                    Return 0

                Catch ex As Exception
                    Trace.WriteLine("A reboot may be required to complete driver setup.")
                    Return 1

                End Try

            Case OpMode.Uninstall
                If AdapterDevicePresent Then
                    Uninstall(ownerWindow)
                    Trace.WriteLine("Driver successfully uninstalled.")

                    Try
                        Using New ScsiAdapter
                        End Using

                        Trace.WriteLine("A reboot may be required to complete driver setup.")
                        Return 2

                    Catch
                        Return 0

                    End Try
                Else
                    Trace.WriteLine("Virtual SCSI adapter not installed.")
                    Return 1
                End If

            Case OpMode.Status
                If AdapterDevicePresent Then
                    Trace.WriteLine("Virtual SCSI adapter installed.")
                    Return 0
                Else
                    Trace.WriteLine("Virtual SCSI adapter not installed.")
                    Return 1
                End If

            Case Else
                Return -1

        End Select

    End Function

End Module

Public Enum OpMode
    Install
    Uninstall
    Status
End Enum

File: Managed Source/ArsenalImageMounterCLISetup/MainModule.cs
using Arsenal.ImageMounter.IO.Native;
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Diagnostics;
using System.Reflection;
using static Arsenal.ImageMounter.API;
// '''' MainModule.vb
// '''' Console driver setup application, for scripting and similar.
// '''' 
// '''' Copyright (c) 2012-2022, Arsenal Consulting, Inc. (d/b/a Arsenal Recon) <http://www.ArsenalRecon.com>
// '''' This source code and API are available under the terms of the Affero General Public
// '''' License v3.
// ''''
// '''' Please see LICENSE.txt for full license terms, including the availability of
// '''' proprietary exceptions.
// '''' Questions, comments, or requests for clarification: http://ArsenalRecon.com/contact/
// ''''

using static Arsenal.ImageMounter.DriverSetup;

#pragma warning disable CS1591 // Missing XML comment for publicly visible type or member

namespace Arsenal.ImageMounter;

public static class MainModule
{

    private static readonly Dialogs.NativeWindowHandle ownerWindow = new(NativeFileIO.SafeNativeMethods.GetConsoleWindow());

    public static int Main(params string[] args)
    {

        Trace.Listeners.Add(new ConsoleTraceListener());

        var opMode = OpMode.Status;

        if (args is not null && args.Length > 0)
        {
            if (args[0].Equals("/install", StringComparison.OrdinalIgnoreCase))
            {
                opMode = OpMode.Install;
            }
            else if (args[0].Equals("/uninstall", StringComparison.OrdinalIgnoreCase))
            {
                opMode = OpMode.Uninstall;
            }
            else if (args[0].Equals("/status", StringComparison.OrdinalIgnoreCase))
            {
                opMode = OpMode.Status;
            }
            else
            {
                Trace.WriteLine($"Syntax: {Assembly.GetExecutingAssembly().GetName().Name} /install|/uninstall|/status");
            }
        }

        try
        {
            return SetupOperation(opMode);
        }
        catch (Exception ex)
        {
            if (ex is Win32Exception win32exception)
            {
                Trace.WriteLine($"Win32 error: {win32exception.NativeErrorCode}");
            }

            Trace.WriteLine(ex.ToString());
            return -1;
        }
    }

    public static int SetupOperation(OpMode opMode)
    {

        Trace.WriteLine($"Kernel type: {Kernel}");
        Trace.WriteLine($"Kernel supports StorPort: {HasStorPort}");

        switch (opMode)
        {

            case OpMode.Install:
                {
                    using (var zipStream = typeof(MainModule).Assembly.GetManifestResourceStream(typeof(MainModule), "DriverFiles.zip")
                        ?? throw new KeyNotFoundException("Cannot find embedded DriverFiles.zip"))
                    {
                        InstallFromZipStream(ownerWindow, zipStream);
                    }

                    try
                    {
                        using (new ScsiAdapter())
                        {
                        }

                        Trace.WriteLine("Driver successfully installed.");
                        return 0;
                    }
                    catch
                    {
                        Trace.WriteLine("A reboot may be required to complete driver setup.");
                        return 1;
                    }
                }

            case OpMode.Uninstall:
                {
                    if (AdapterDevicePresent)
                    {
                        Uninstall(ownerWindow);
                        Trace.WriteLine("Driver successfully uninstalled.");

                        try
                        {
                            using (new ScsiAdapter())
                            {
                            }

                            Trace.WriteLine("A reboot may be required to complete driver setup.");
                            return 2;
                        }
                        catch
                        {
                            return 0;
                        }
                    }
                    else
                    {
                        Trace.WriteLine("Virtual SCSI adapter not installed.");
                        return 1;
                    }
                }

            case OpMode.Status:
                {
                    if (AdapterDevicePresent)
                    {
                        Trace.WriteLine("Virtual SCSI adapter installed.");
                        return 0;
                    }
                    else
                    {
                        Trace.WriteLine("Virtual SCSI adapter not installed.");
                        return 1;
                    }
                }

            default:
                {
                    return -1;
                }
        }
    }
}

public enum OpMode
{
    Install,
    Uninstall,
    Status
}

